### PR TITLE
1023 Add default function for user-constructing collection elements

### LINKED ISSUE #1023: Add default function for user-constructing collection elements
**What Needs to be Done?**

Right now, even if a `vt::Collection` is default constructible, the user has to provide a lambda that calls `std::make_unique<ColT>()` to construct collection elements. Add a default function that just calls `make_unique` when nothing is passed:

Example from hello_world_collection_collective.cc
https://github.com/DARMA-tasking/vt/blob/f9b7e74a5f497a94687d6c3a97349f8908ce963d/examples/hello_world/hello_world_collection_collective.cc#L75-L78

Ideally, this would be allowed:
```C++
  auto range = vt::Index1D(num_elms);
  auto proxy = vt::theCollection()->constructCollective<Hello>(range);
```

This might also need to change in some of the overloads possibly for `construct` also.

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #1023

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/collection/test_lb.extended.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6292206&bid=20470119)

File: src/vt/vrt/collection/manager.h [hunk 1]
```diff
@@ -222,6 +227,27 @@ struct CollectionManager
...
+  /**
+   * \brief Collectively construct a new virtual context collection with
+   * templated map
+   *
+   *  Construct virtual context collection with an explicit map. This construct
+   *  method enables distributed SPMD construction of the virtual context
+   *  collection where each index is mapped with the \c MapFnT.
+   *
+   * \param[in] range index range for the collection
+   * \param[in] cons_fn construct function to create an element on each node
```
Comment (PhilMiller) [inline src/vt/vrt/collection/manager.h:239]: Copy-paste mistake here - there is no `cons_fn`, since that's the point of this overload.

**Review by PhilMiller — APPROVED** [review]
Other than the comment typo, I think this is fine to merge.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/collection/test_lb.extended.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6292206&bid=20470119)

File: src/vt/vrt/collection/manager.h (hunk 1 above)
Comment (JacobDomagala) [reply src/vt/vrt/collection/manager.h]: Nice catch, thanks

File: src/vt/vrt/collection/manager.impl.h [hunk 2]
```diff
@@ -1466,6 +1466,16 @@ bool CollectionManager::insertCollectionElement(
  * user supplying the pointers to the elements.
  */
 
+template <typename ColT>
+CollectionManager::IsDefaultConstructableType<ColT>
+ CollectionManager::constructCollective(
```
Comment (lifflander) [inline src/vt/vrt/collection/manager.impl.h:1471]: Indentation might be slightly off here.

**Review by lifflander — APPROVED** [review]
Looks good to me

File: src/vt/vrt/collection/manager.impl.h (hunk 2 above)
Comment (JacobDomagala) [reply src/vt/vrt/collection/manager.impl.h]: Fixed

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/collection/test_lb.extended.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6292206&bid=20470119)

**Pull request merged**